### PR TITLE
Add rem builtin for i32 and i64 (#1310)

### PR DESCRIPTION
**Pull request by adrien-zinger**
Implement the builtin `rem` for `i32` and `i64`, requested in the issue #1310 

- [x] I've read the contributing guidelines
- [x] I've added my name and email to the NOTICE file

### CODE REVIEW

**Comment by MaxGraey** [comment]
Hi! Thanks for PR. Similar to division builtins it should be splitted to rem_s and rem_u. See:
https://github.com/AssemblyScript/assemblyscript/blob/a2acee0df5012dfd3e3c6b4b82ad98a75fc45bb0/src/builtins.ts#L6826
https://github.com/AssemblyScript/assemblyscript/blob/a2acee0df5012dfd3e3c6b4b82ad98a75fc45bb0/src/builtins.ts#L6835

**Comment by adrien-zinger** [comment]
Thanks for the review @MaxGraey! I just implemented the split. I have a doubt on types in the std because the `div_s` and `div_u` have both a `i32/64` type for the parameters. It should be `u32/64` right? Even if the compiler will reinterpret those types later (if I understand correctly :smile: ).

**Comment by MaxGraey** [comment]
plz update test/features as well. Just run this command:
```bash
ASC_FEATURES="*" npm run test:compiler -- --create
```

File: std/assembly/index.d.ts [hunk 1]
```diff
@@ -151,6 +151,8 @@ declare function div<T extends i32 | i64 | f32 | f64>(left: T, right: T): T;
 declare function eq<T extends i32 | i64 | f32 | f64>(left: T, right: T): i32;
 /** Return 0 if two numbers are equal to each other, 1 otherwise. */
 declare function ne<T extends i32 | i64 | f32 | f64>(left: T, right: T): i32;
+/** Computes the rem of two positive integers. */
```
Comment (dcodeIO) [inline std/assembly/index.d.ts:154]:
```suggestion
/** Computes the remainder of two integers. */
```

File: std/assembly/index.d.ts [hunk 2]
```diff
@@ -342,10 +344,15 @@ declare namespace i32 {
...
-  /** Return 1 two 32-bit inegers are equal to each other, 0 otherwise. */
+  /** Return 1 if two 32-bit integers are equal to each other, 0 otherwise. */
   export function eq(left: i32, right: i32): i32;
-  /** Return 0 two 32-bit inegers are equal to each other, 1 otherwise. */
+  /** Return 0 if two 32-bit integers are equal to each other, 1 otherwise. */
   export function ne(left: i32, right: i32): i32;
+  /** Compute the remaining between two 32-bit signed positive integers. */
+  export function rem_s(left: i32, right: i32): i32;
+  /** Compute the remaining between two 32-bit unsigned positive integers. */
+  export function rem_u(left: u32, right: u32): u32;
```
Comment (dcodeIO) [inline std/assembly/index.d.ts:354]:
```suggestion
  /** Computes the signed remainder of two 32-bit integers. */
  export function rem_s(left: i32, right: i32): i32;
  /** Computes the unsigned remainder of two 32-bit integers. */
  export function rem_u(left: u32, right: u32): u32;
```

File: std/assembly/index.d.ts [hunk 3]
```diff
@@ -466,10 +473,15 @@ declare namespace i64 {
...
-  /** Return 1 two 64-bit inegers are equal to each other, 0 otherwise. */
+  /** Return 1 if two 64-bit integers are equal to each other, 0 otherwise. */
   export function eq(left: i64, right: i64): i32;
-  /** Return 0 two 64-bit inegers are equal to each other, 1 otherwise. */
+  /** Return 0 if two 64-bit integers are equal to each other, 1 otherwise. */
   export function ne(left: i64, right: i64): i32;
+  /** Compute the remaining between two 64-bit signed positive integers. */
+  export function rem_s(left: i64, right: i64): i64;
+  /** Compute the remaining between two 64-bit unsigned positive integers. */
+  export function rem_u(left: u64, right: u64): u64;
```
Comment (dcodeIO) [inline std/assembly/index.d.ts:483]:
```suggestion
  /** Computes the signed remainder of two 64-bit integers. */
  export function rem_s(left: i64, right: i64): i64;
  /** Computes the unsigned remainder of two 64-bit integers. */
  export function rem_u(left: u64, right: u64): u64;
```

**Comment by dcodeIO** [comment]
Thanks, looks good, just a few nits on wording above :)

Regarding

> I have a doubt on types in the std because the `div_s` and `div_u` have both a `i32/64` type for the parameters. It should be `u32/64` right? Even if the compiler will reinterpret those types later (if I understand correctly 😄 ).

In the `.d.ts`, the numeric types are merely aliases of `number`, so I guess that's why the discrepancy hasn't been noticed before. Doesn't harm to unify these, as you suggest. Perhaps as another PR?

File: std/assembly/index.d.ts (hunk 3 above)
Comment (adrien-zinger) [reply std/assembly/index.d.ts]: done in https://github.com/AssemblyScript/assemblyscript/pull/2306/commits/1130f0aac50bf058c79d91228484c4e872b53771

File: std/assembly/index.d.ts (hunk 2 above)
Comment (adrien-zinger) [reply std/assembly/index.d.ts]: done in https://github.com/AssemblyScript/assemblyscript/pull/2306/commits/1130f0aac50bf058c79d91228484c4e872b53771

File: std/assembly/index.d.ts (hunk 1 above)
Comment (adrien-zinger) [reply std/assembly/index.d.ts]: done in https://github.com/AssemblyScript/assemblyscript/pull/2306/commits/1130f0aac50bf058c79d91228484c4e872b53771

**Comment by adrien-zinger** [comment]
> In the `.d.ts`, the numeric types are merely aliases of `number`, so I guess that's why the discrepancy hasn't been noticed before. Doesn't harm to unify these, as you suggest. Perhaps as another PR?

Thanks for your review! I can change that in another PR as you said :+1:

**Review by dcodeIO — APPROVED** [review]

**Pull request merged**